### ISSUE TITLE
test plugins::openshift_secondary_metadata_parser::plugin::tests::compare_quay_result_fixture_20200220104838 ... FAILED

### ISSUE THREAD

**Issue by admiyo** [openshift/cincinnati#236]
Plugin test failure
From git commit  7f68cf5ceadd4cee22b7cc8ae319097954c860e9

cargo clean; cargo build; cargo test
...
failures:

---- plugins::openshift_secondary_metadata_parser::plugin::tests::compare_quay_result_fixture_20200220104838 stdout ----
thread 'plugins::openshift_secondary_metadata_parser::plugin::tests::compare_quay_result_fixture_20200220104838' panicked at 'assertion failed: `(left == right)`

**Comment by admiyo** [comment]
The diff is here:

                                 "io.openshift.upgrades.graph.release.channels": String(
<                                    "stable-4.3,candidate-4.3,fast-4.3",
>                                    "candidate-4.3,fast-4.3,stable-4.3",
                                 ),
Order of the elements of the string has changed.

**Comment by steveej** [comment]
I'm surprised as I can't reproduce this and I couldn't explain how this would have passed through CI. Are you sure you're on a clean working tree?

**Comment by admiyo** [comment]
Yes, just reconfirmed from a clone.  What is happening to generates the lines:"io.openshift.upgrades.graph.release.channels": String(

If it is something where order is not guaranteed, like a hashmap, it might be a version difference.

**Comment by admiyo** [comment]
The string is built by the closure run here: 

https://github.com/openshift/cincinnati/blob/master/graph-builder/src/plugins/openshift_secondary_metadata_parser/plugin.rs#L379

Which compares the second half of the string (part after the - in the  a_split[1].cmp(b_split[1]))  

The original code it attempts to duplicate is (in python)

            channels = sorted(node['channels'])
            # then sort by version, 4.1, 4.2, etc
            channels = sorted(channels, key=lambda x: x.split('-', 1)[-1])

If the two values are equal, it seems to be non-determinisition which value would be returned.  I think the intention is to leave the first part alone.

I have a fix that works on my machine that performs that first sort.

**Comment by steveej** [comment]
I'm strangely still not able to reproduce the issue as you can see here:

```consoles
steveej@steveej-t480s-work ✓ ~/src/job-redhat/cincinnati/graph-builder  7f68cf5 ⚡
$ (
  set -e
  for i in `seq 1 100`; do
    cargo test compare_quay_result_fixture_20200220104838 2>/dev/null
  done
) | rg -c 'compare_quay_result_fixture_20200220104838 ... ok'
100
```
 Did you modify the test before running it in any way?

I'm fine merging the linked PR anyway, with or without the suggestion I've made on it, as it does match the python version more accurately which was the goal.

**Comment by admiyo** [comment]
I suspect that there is something going on in the rust compiler or libraries.  I made no changes other than adding the single line.

The data getting sorted comes from here: 

let mut channels_split = channels.split(',').collect::<Vec<_>>();

So if either the **split** or the **collect** call changes the order it uses between versions of rust, we'll see differences.  I used:

$ rustc --version
rustc 1.41.1 (f3e1a954d 2020-02-24)

It might be earlier where the **channels** variable is created.  Hard to track the logic back to that point.  I do see that there are lots of hashmaps used, and hashmaps do not maintain order.
